gitlabform/processors/project/schedules_processor.py

```diff
@@ -42,102 +48,124 @@ def _process_configuration(self, project_and_group: str, configuration: dict):
+            # Delete and then re-create schedule so we can pass all info in the data to Gitlab in case their APIs change
+            project.pipelineschedules.delete(schedule.id)
 
-                debug(f"processing {schedule_id}: {schedule_description}")
-                if schedule_description not in configured_schedules:
-                    debug(
-                        "Deleting pipeline schedule named '%s', because it is not in gitlabform configuration",
-                        schedule_description,
-                    )
-                    self.gitlab.delete_pipeline_schedule(project_and_group, schedule_id)
+            self._create_schedule_with_variables(
+                configured_schedules, project, schedule_description
+            )
 
-    def create_schedule_with_variables(
-        self, configuration, project_and_group, schedule_description
+    def _create_schedule_with_variables(
+        self, configured_schedules, project: Project, schedule_description: str
     ):
-        data = configuration.get("schedules|" + schedule_description)
-        created_schedule = self.gitlab.create_pipeline_schedule(
-            project_and_group,
-            schedule_description,
-            data.get("ref"),
-            data.get("cron"),
-            optional_data=data,
-        )
-        self.__set_schedule_variables(
-            project_and_group,
-            created_schedule.get("id"),
-            configuration.get("schedules|" + schedule_description + "|variables"),
+        data = configured_schedules[schedule_description]
+
+        schedule_data = {"description": schedule_description, **data}
+        debug("Creating pipeline schedule using data: '%s'", schedule_data)
+
+        created_schedule_id = project.pipelineschedules.create(schedule_data).id
+        created_schedule = project.pipelineschedules.get(created_schedule_id)
+
+        self._set_schedule_variables(
+            created_schedule,
+            data.get("variables"),
         )
 
-    def __set_schedule_variables(self, project_and_group, schedule_id, variables):
-        schedule = self.gitlab.get_pipeline_schedule(project_and_group, schedule_id)
+        created_schedule.save()
 
-        existing_variables = schedule.get("variables")
+    @staticmethod
+    def _set_schedule_variables(schedule: ProjectPipelineSchedule, variables):
+        attributes = schedule.attributes
+        existing_variables = attributes.get("variables")
         if existing_variables:
-            debug(
-                "Deleting variables for pipeline schedule '%s'", schedule["description"]
-            )
+            debug("Deleting variables for pipeline schedule '%s'", schedule.description)
 
             for variable in existing_variables:
-                self.gitlab.delete_pipeline_schedule_variable(
-                    project_and_group, schedule_id, variable.get("key")
-                )
+                schedule.variables.delete(variable.get("key"))
```

Comment:
I know deleting the existing variable was being done before but I wonder if that's really needed. Going through the flow of code in your change, the `_set_schedule_variables` being called only when a pipeline is being created. So, how would there be existing schedule?

What do you think?

Reply:
Hmm yeah even in the current update flow, the old schedule gets deleted and a new one created, so this isn't needed. If I get update working without delete/create will need to call _set_schedule_variables from there

Reply:
@amimas okay this is all done now - there are definetly problems with the test suite making it hard to add new test cases. They all modify the same project and don't clean up pipelineschedules at the end of each test run, so there's lots of awkward interactions when running the full suite vs running a test individually.

Probably need to raise an issue to fully re-write the tests, I don't have the brain to do it at the moment (spent about 4hrs today trying to unpick what was happening)

Reply:
> even in the current update flow, the old schedule gets deleted and a new one created

Not sure if we're looking at the same thing but I believe the above happens in a very specific scenario. You'll notice the schedules are first grouped by their descriptions. If there are multiple schedules with the same description, then the above happens (i.e. delete all these schedules and create a new single schedule with given description). Otherwise I believe the delete and create new does not happen right now.

> They all modify the same project and don't clean up pipelineschedules at the end of each test run, so there's lots of awkward interactions when running the full suite vs running a test individually.

Yeah, I know what you're referring to. I think it has both pros and cons. Sometimes that setup is useful and probably makes subsequent tests faster and easier. But sometimes that does get in the way.

You'll notice the tests are using `project` fixture. This is scoped to the entire test class. That's why the same project is available to every test case/function. There's another fixture named `project_for_function`. Sounds like this is probably what you were looking for. It's being used in other test cases where needed.

Reply:
@TimKnight-Opencast - I didn't review the updates to the test cases. Do you want to refactor your test updates using the `project_for_function` fixture if needed?